src/core/thread/mle.cpp

```diff
@@ -4743,5 +4769,72 @@ void Mle::DelayedResponseMetadata::RemoveFrom(Message &aMessage) const
     SuccessOrAssert(aMessage.SetLength(aMessage.GetLength() - sizeof(*this)));
 }
 
+Error Mle::DetachGracefully(otDetachGracefullyCallback aCallback, void *aContext)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(!IsDetachingGracefully(), error = kErrorBusy);
+
+    OT_ASSERT(mDetachGracefullyCallback == nullptr);
+
+    mDetachGracefullyCallback = aCallback;
+    mDetachGracefullyContext  = aContext;
+
+    if (IsChild() || IsRouter())
```

Comment:
We can be a leader as well (`IsChild()`, `IsRouter()`, `IsLeader()` are just helpers for  checking `GetRole()` against different constants). 
```suggestion
    if (IsChild() || IsRouterOrLeader())
```

Reply:
IIUC I don't think we should include leader here because a leader can't really do anything. Sending an address release to itself followed by detaching doesn't seem to make sense. So I intentionally made the leader case the same as detached or disabled.

Reply:
I see. Yes. Makes sense. Thanks.